mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,328 @@
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+from pytools.obj_array import obj_array_vectorize
+from meshmode.dof_array import thaw, DOFArray
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+from grudge.eager import interior_trace_pair, cross_rank_trace_pairs
+from grudge.symbolic.primitives import TracePair
+from grudge.dof_desc import DD_VOLUME_MODAL, DD_VOLUME
+
+
+def _facial_flux_q(discr, q_tpair):
+    """Compute facial flux for each scalar component of Q."""
+    flux_dis = q_tpair.avg
+    if isinstance(flux_dis, np.ndarray):
+        actx = flux_dis[0].array_context
+        flux_dis = flux_dis.reshape(-1, 1)
+    else:
+        actx = flux_dis.array_context
+
+    normal = thaw(actx, discr.normal(q_tpair.dd))
+
+    # This uses a central scalar flux along nhat:
+    # flux = 1/2 * (Q- + Q+) * nhat
+    flux_out = flux_dis * normal
+
+    return discr.project(q_tpair.dd, "all_faces", flux_out)
+
+
+def _facial_flux_r(discr, r_tpair):
+    """Compute facial flux for vector component of grad(Q)."""
+    flux_dis = r_tpair.avg
+    if isinstance(flux_dis[0], np.ndarray):
+        actx = flux_dis[0][0].array_context
+    else:
+        actx = flux_dis[0].array_context
+
+    normal = thaw(actx, discr.normal(r_tpair.dd))
+
+    # This uses a central vector flux along nhat:
+    # flux = 1/2 * (grad(Q)- + grad(Q)+) .dot. nhat
+    flux_out = flux_dis @ normal
+
+    return discr.project(r_tpair.dd, "all_faces", flux_out)
+
+
+def av_operator(discr, t, eos, boundaries, q, alpha, **kwargs):
+    r"""Compute the artificial viscosity right-hand-side.
+
+    Computes the the right-hand-side term for artificial viscosity.
+
+    .. math::
+
+        \mbox{RHS}_{\mbox{av}} = \nabla\cdot{\varepsilon\nabla\mathbf{Q}}
+
+    Parameters
+    ----------
+    q: :class:`~meshmode.dof_array.DOFArray` or :class:`~numpy.ndarray`
+
+        Single :class:`~meshmode.dof_array.DOFArray` for a scalar or an object array
+        (:class:`~numpy.ndarray`) for a vector of
+        :class:`~meshmode.dof_array.DOFArray` on which to operate.
+
+        When used with fluid solvers, *q* is expected to be the fluid state array
+        of the canonical conserved variables (mass, energy, momentum)
+        for the fluid along with a vector of species masses for multi-component
+        fluids.
+
+    boundaries: float
+
+        Dictionary of boundary functions, one for each valid boundary tag
+
+    t: float
+
+        Time
+
+    alpha: float
+
+       The maximum artificial viscosity coefficient to be applied
+
+    eos: :class:`~mirgecom.eos.GasEOS`
+
+       Only used as a pass through to the boundary conditions.
```

Comment:
Would it make sense to pass something like a `**bdry_kwargs` instead of passing the EOS explicitly?

Reply:
How's come it couldn't be passed through the **kwargs?  (assuming you are suggesting the boundary should have its own kwargs struct).

Reply:
I tried doing something like that once and I encountered some problems. That said I don't recall what they were (probably me not understanding python...). I can give something like this another  try though.

Reply:
Should time (*t*) also go into **kwargs, or whatever is used to contain args to boundary routines?

Reply:
w-hagen/mirgecom#23